File: LeetCode/LC.3325.count-substrings-with-k-frequency-characters-i.cpp
#include "../utils/abel_macro.h"

class Solution {
public:
    int numberOfSubstrings(string s, int k) {
        int ans = 0, left = 0, cnt[26]{};
        for (char c : s) {
            cnt[c-'a']++;
            while (cnt[c-'a'] >= k) {
                cnt[s[left]-'a']--;
                left++;
            }
            ans += left;
        }
        return ans;
    }
};

File: LeetCode/LC.3325.count-substrings-with-k-frequency-characters-i.py
from lc_utils import *


class Solution:
    def numberOfSubstrings(self, s: str, k: int) -> int:
        cnt = Counter()
        valid = 0
        ans = l = 0
        for c in s:
            cnt[c] += 1
            if cnt[c] == k:
                valid += 1
            while valid:
                out = s[l]
                if cnt[out] == k:
                    valid -= 1
                cnt[out] -= 1
                l += 1
            ans += l
        return ans
